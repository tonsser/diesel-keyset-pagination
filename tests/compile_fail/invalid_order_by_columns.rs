#![allow(unused_imports)]

#[macro_use]
extern crate diesel;

use chrono::prelude::*;
use diesel_keyset_pagination::*;
use diesel::prelude::*;
use schema::*;

mod schema {
    table! {
        users (id) {
            firstname -> Text,
            id -> Integer,
            lastname -> Text,
            slug -> Text,
        }
    }

    table! {
        follows (id) {
            followee_id -> Integer,
            followee_type -> Nullable<Text>,
            follower_id -> Integer,
            id -> Integer,
            source -> Nullable<Text>,
            unfollowed_at -> Nullable<Timestamptz>,
        }
    }
}

#[derive(Eq, PartialEq, Debug, Clone, Queryable)]
pub struct User {
    pub firstname: String,
    pub id: i32,
    pub lastname: String,
    pub slug: String,
}

fn main() {
    let url = "postgres://localhost/tonsser-api_test";
    let db = PgConnection::establish(url).unwrap();

    users::table
        .select(users::all_columns)
        .keyset_paginate_order_by(follows::id)
        .page_size(2)
        .cursor(users::id, None::<i32>)
        .load::<User>(&db)
        .unwrap();
}
