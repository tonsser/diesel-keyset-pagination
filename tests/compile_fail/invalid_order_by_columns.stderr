error[E0277]: the trait bound `diesel::query_builder::SelectStatement<schema::users::table, diesel::query_builder::select_clause::SelectClause<(schema::users::columns::firstname, schema::users::columns::id, schema::users::columns::lastname, schema::users::columns::slug)>>: diesel::query_dsl::select_dsl::SelectDsl<schema::follows::columns::id>` is not satisfied
  --> $DIR/invalid_order_by_columns.rs:50:10
   |
50 |         .load::<User>(&db)
   |          ^^^^ the trait `diesel::query_dsl::select_dsl::SelectDsl<schema::follows::columns::id>` is not implemented for `diesel::query_builder::SelectStatement<schema::users::table, diesel::query_builder::select_clause::SelectClause<(schema::users::columns::firstname, schema::users::columns::id, schema::users::columns::lastname, schema::users::columns::slug)>>`
   |
   = help: the following implementations were found:
             <diesel::query_builder::SelectStatement<F, S, D, W, O, L, Of, G, LC> as diesel::query_dsl::select_dsl::SelectDsl<Selection>>
   = note: required because of the requirements on the impl of `diesel::query_builder::QueryFragment<diesel::pg::Pg>` for `diesel_keyset_pagination::KeysetPaginated<diesel::query_builder::SelectStatement<schema::users::table, diesel::query_builder::select_clause::SelectClause<(schema::users::columns::firstname, schema::users::columns::id, schema::users::columns::lastname, schema::users::columns::slug)>>, schema::follows::columns::id, i32, schema::users::columns::id>`
   = note: required because of the requirements on the impl of `diesel::query_dsl::LoadQuery<_, User>` for `diesel_keyset_pagination::KeysetPaginated<diesel::query_builder::SelectStatement<schema::users::table, diesel::query_builder::select_clause::SelectClause<(schema::users::columns::firstname, schema::users::columns::id, schema::users::columns::lastname, schema::users::columns::slug)>>, schema::follows::columns::id, i32, schema::users::columns::id>`

For more information about this error, try `rustc --explain E0277`.
